README.md
# viridisLite - Colorblind-Friendly Color Maps for R <img src="man/figures/logo.png" align="right" alt="" width="120" />

[![cran version](http://www.r-pkg.org/badges/version/viridisLite)](https://cran.r-project.org/package=viridisLite)
[![rstudio mirror per-month downloads](http://cranlogs.r-pkg.org/badges/viridisLite)](https://github.com/metacran/cranlogs.app)
[![rstudio mirror total downloads](http://cranlogs.r-pkg.org/badges/grand-total/viridisLite?color=yellowgreen)](https://github.com/metacran/cranlogs.app)
[![R-CMD-check](https://github.com/sjmgarnier/viridisLite/workflows/R-CMD-check/badge.svg)](https://github.com/sjmgarnier/viridisLite/actions)
[![Coverage Status](https://img.shields.io/codecov/c/github/sjmgarnier/viridisLite/master.svg)](https://codecov.io/github/sjmgarnier/viridisLite?branch=master)
[![DOI](https://zenodo.org/badge/44360137.svg)](https://zenodo.org/badge/latestdoi/44360137)

## Description

`viridisLite`, and its companion package [`viridis`](https://cran.r-project.org/package=viridis) 
provide a series of color maps that are designed to improve graph readability 
for readers with common forms of color blindness and/or color vision deficiency. 
The color maps are also perceptually-uniform, both in regular form and also when 
converted to black-and-white for printing. 

`viridisLite` provides the base functions for generating the color maps in base 
`R`. The package is meant to be as lightweight and dependency-free as possible 
for maximum compatibility with all the `R` ecosystem. [`viridis`](https://cran.r-project.org/package=viridis)
provides additional functionalities, in particular bindings for `ggplot2`.

---

## The color maps

The latest version of `viridisLite` comes with 8 different color maps, and they 
are all very pretty!!!

![Sample image](man/figures/maps.png)

---

## Installation

You can install `viridisLite` from `CRAN` by typing the following line in your R 
console:

```{r}
install.packages("viridisLite")
library(viridisLite)
```

If you prefer to install the development version from this GitHub repository,
simply copy the following lines of code in your R terminal and it should install 
everything you need to use `viridisLite` on your computer: 

```{r}
if (!require("devtools")) 
  install.packages("devtools")

devtools::install_github("sjmgarnier/viridisLite")
library(viridisLite)
```

---

## References

The color maps `viridis`, `magma`, `inferno`, and `plasma` were created by 
Stéfan van der Walt ([@stefanv](https://github.com/stefanv)) and Nathaniel Smith ([@njsmith](https://github.com/njsmith)). If you want to know more about the 
science behind the creation of these color maps, you can watch this 
[presentation of `viridis`](https://youtu.be/xAoljeRJ3lU) by their authors at 
[SciPy 2015](http://scipy2015.scipy.org/). 

The color map `cividis` is a corrected version of 'viridis', developed by 
Jamie R. Nuñez, Christopher R. Anderton, and Ryan S. Renslow, and originally 
ported to `R` by Marco Sciaini ([@msciain](https://github.com/marcosci)). More 
info about `cividis` can be found in 
[this paper](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0199239).

The color maps `mako` and `rocket` were originally created for the `Seaborn` 
statistical data visualization package for Python. More info about `mako` and 
`rocket` can be found on the 
[`Seaborn` website](https://seaborn.pydata.org/tutorial/color_palettes.html).

The color map `turbo` was developed by Anton Mikhailov to address the 
shortcomings of the Jet rainbow color map such as false detail, banding and 
color blindness ambiguity. More infor about `turbo` can be found
[here](https://ai.googleblog.com/2019/08/turbo-improved-rainbow-colormap-for.html).


NEWS.md
# viridisLite 0.4.1

## New features

* N/A.

## Minor improvements and fixes

* Fix CRAN check warnings with R-devel. 

---

# viridisLite 0.4.0

## New features

* Add 3 more color maps: mako, rocket, and turbo. 

## Minor improvements and fixes

* Minor bug fixes and improvements here and there. 

---

tests/testthat/test-palettes.R
context("palette generation")
test_that("palette generation is accurate", {

  # data is intact
  expect_equal(dim(viridis.map), c(2048, 4))

  # options work as expected
  expect_equal(viridis(1, option = "D"), "#440154FF")
  expect_equal(viridis(1, alpha = 0, option = "D"), "#44015400")
  expect_equal(viridis(1, begin = 1, end = 0, option = "D"), "#FDE725FF")
  expect_equal(viridis(1, alpha = 0, begin = 1, end = 0, option = "D"), "#FDE72500")
  expect_equal(viridis(2, alpha = 0, begin = 1, end = 0, direction = -1, option = "D"),
               c("#44015400", "#FDE72500"))

  # options work as expected
  expect_equal(viridis(1, option = "C"), "#0D0887FF")
  expect_equal(viridis(1, alpha = 0, option = "C"), "#0D088700")
  expect_equal(viridis(1, begin = 1, end = 0, option = "C"), "#F0F921FF")
  expect_equal(viridis(1, alpha = 0, begin = 1, end = 0, option = "C"), "#F0F92100")
  expect_equal(viridis(2, alpha = 0, begin = 1, end = 0, direction = -1, option = "C"),
               c("#0D088700", "#F0F92100"))

  # options work as expected
  expect_equal(viridis(1, option = "B"), "#000004FF")
  expect_equal(viridis(1, alpha = 0, option = "B"), "#00000400")
  expect_equal(viridis(1, begin = 1, end = 0, option = "B"), "#FCFFA4FF")
  expect_equal(viridis(1, alpha = 0, begin = 1, end = 0, option = "B"), "#FCFFA400")
  expect_equal(viridis(2, alpha = 0, begin = 1, end = 0, direction = -1, option = "B"),
               c("#00000400", "#FCFFA400"))

  # options work as expected
  expect_equal(viridis(1, option = "A"), "#000004FF")
  expect_equal(viridis(1, alpha = 0, option = "A"), "#00000400")
  expect_equal(viridis(1, begin = 1, end = 0, option = "A"), "#FCFDBFFF")
  expect_equal(viridis(1, alpha = 0, begin = 1, end = 0, option = "A"), "#FCFDBF00")
  expect_equal(viridis(2, alpha = 0, begin = 1, end = 0, direction = -1, option = "A"),
               c("#00000400", "#FCFDBF00"))

  # options work as expected
  expect_equal(viridis(1, option = "E"), "#00204DFF")
  expect_equal(viridis(1, alpha = 0, option = "E"), "#00204D00")
  expect_equal(viridis(1, begin = 1, end = 0, option = "E"), "#FFEA46FF")
  expect_equal(viridis(1, alpha = 0, begin = 1, end = 0, option = "E"), "#FFEA4600")
  expect_equal(viridis(2, alpha = 0, begin = 1, end = 0, direction = -1, option = "E"),
              c("#00204D00", "#FFEA4600"))

  # options work as expected
  expect_equal(viridis(1, option = "F"), "#03051AFF")
  expect_equal(viridis(1, alpha = 0, option = "F"), "#03051A00")
  expect_equal(viridis(1, begin = 1, end = 0, option = "F"), "#FAEBDDFF")
  expect_equal(viridis(1, alpha = 0, begin = 1, end = 0, option = "F"), "#FAEBDD00")
  expect_equal(viridis(2, alpha = 0, begin = 1, end = 0, direction = -1, option = "F"),
              c("#03051A00", "#FAEBDD00"))

  # options work as expected
  expect_equal(viridis(1, option = "G"), "#0B0405FF")
  expect_equal(viridis(1, alpha = 0, option = "G"), "#0B040500")
  expect_equal(viridis(1, begin = 1, end = 0, option = "G"), "#DEF5E5FF")
  expect_equal(viridis(1, alpha = 0, begin = 1, end = 0, option = "G"), "#DEF5E500")
  expect_equal(viridis(2, alpha = 0, begin = 1, end = 0, direction = -1, option = "G"),
              c("#0B040500", "#DEF5E500"))

  # options work as expected
  expect_equal(viridis(1, option = "H"), "#30123BFF")
  expect_equal(viridis(1, alpha = 0, option = "H"), "#30123B00")
  expect_equal(viridis(1, begin = 1, end = 0, option = "H"), "#7A0403FF")
  expect_equal(viridis(1, alpha = 0, begin = 1, end = 0, option = "H"), "#7A040300")
  expect_equal(viridis(2, alpha = 0, begin = 1, end = 0, direction = -1, option = "H"),
               c("#30123B00", "#7A040300"))

  # bad inputs
  expect_warning(viridis(1, option = "I"))
  expect_error(viridis(1, direction = 100))
  expect_error(viridis(1, begin = -1))
  expect_error(viridis(1, begin = 100))
  expect_error(viridis(1, end = -1))
  expect_error(viridis(1, end = 100))

  # zero input, zero output
  expect_equal(viridis(0), character(0))
  expect_equal(mako(0), character(0))
  expect_equal(viridis(0, begin = 0.9, end = 0.2), character(0))
  expect_equal(nrow(viridisMap(0)), 0)

  # we've already proven these work with ^^ but we'll add a few more values to
  # generate and get better code coverage this way.

  expect_equal(magma(3), c("#000004FF", "#B63679FF", "#FCFDBFFF"))
  expect_equal(inferno(3), c("#000004FF", "#BB3754FF", "#FCFFA4FF"))
  expect_equal(plasma(3), c("#0D0887FF", "#CC4678FF", "#F0F921FF"))
  expect_equal(cividis(3), c("#00204DFF", "#7C7B78FF", "#FFEA46FF"))
  expect_equal(rocket(3), c("#03051AFF", "#CB1B4FFF", "#FAEBDDFF"))
  expect_equal(mako(3), c("#0B0405FF", "#357BA2FF", "#DEF5E5FF"))
  expect_equal(turbo(3), c("#30123BFF", "#A2FC3CFF", "#7A0403FF"))
})


data-raw/maps.R
library(viridisLite)
library(ggplot2)
library(maps)
library(scales)
library(patchwork)

unemp <- read.csv("http://datasets.flowingdata.com/unemployment09.csv",
                  header = FALSE, stringsAsFactors = FALSE)
names(unemp) <- c("id", "state_fips", "county_fips", "name", "year",
                  "?", "?", "?", "rate")
unemp$county <- tolower(gsub(" County, [A-Z]{2}", "", unemp$name))
unemp$county <- gsub("^(.*) parish, ..$","\\1", unemp$county)
unemp$state <- gsub("^.*([A-Z]{2}).*$", "\\1", unemp$name)

county_df <- map_data("county", projection = "albers", parameters = c(39, 45))
names(county_df) <- c("long", "lat", "group", "order", "state_name", "county")
county_df$state <- state.abb[match(county_df$state_name, tolower(state.name))]
county_df$state_name <- NULL

state_df <- map_data("state", projection = "albers", parameters = c(39, 45))

choropleth <- merge(county_df, unemp, by = c("state", "county"))
choropleth <- choropleth[order(choropleth$order), ]

base_graph <- ggplot(choropleth, aes(long, lat, group = group)) +
  geom_polygon(aes(fill = rate), colour = alpha("white", 1 / 2), size = 0.1) +
  geom_polygon(data = state_df, colour = "white", fill = NA, size = 0.25) +
  coord_fixed() +
  theme_minimal() +
  theme(axis.line = element_blank(), axis.text = element_blank(),
        axis.ticks = element_blank(), axis.title = element_blank())

optA <- base_graph +
  scale_fill_gradientn("", colours = viridis(256, option = "A")) +
  ggtitle("option A aka 'magma'") +
  theme(plot.margin = unit(c(0, 0, 0, 0), "cm"))
optB <- base_graph +
  scale_fill_gradientn("", colours = viridis(256, option = "B")) +
  ggtitle("option B aka 'inferno'") +
  theme(plot.margin = unit(c(0, 0, 0, 0), "cm"))
optC <- base_graph +
  scale_fill_gradientn("", colours = viridis(256, option = "C")) +
  ggtitle("option C aka 'plasma'") +
  theme(plot.margin = unit(c(0, 0, 0, 0), "cm"))
optD <- base_graph +
  scale_fill_gradientn("", colours = viridis(256, option = "D")) +
  ggtitle("option D aka 'viridis'") +
  theme(plot.margin = unit(c(0, 0, 0, 0), "cm"))
optE <- base_graph +
  scale_fill_gradientn("", colours = viridis(256, option = "E")) +
  ggtitle("option E aka 'cividis'") +
  theme(plot.margin = unit(c(0, 0, 0, 0), "cm"))
optF <- base_graph +
  scale_fill_gradientn("", colours = viridis(256, option = "F")) +
  ggtitle("option F aka 'rocket'") +
  theme(plot.margin = unit(c(0, 0, 0, 0), "cm"))
optG <- base_graph +
  scale_fill_gradientn("", colours = viridis(256, option = "G")) +
  ggtitle("option G aka 'mako'") +
  theme(plot.margin = unit(c(0, 0, 0, 0), "cm"))
optH <- base_graph +
  scale_fill_gradientn("", colours = viridis(256, option = "H")) +
  ggtitle("option H aka 'turbo'") +
  theme(plot.margin = unit(c(0, 0, 0, 0), "cm"))

png("man/figures/maps.png", width = 1500, height = 1000, res = 150)
(optA + optB + optC) /
  (optD + optE) /
  (optF + optG + optH)
dev.off()


data-raw/viridis_map.R
optionA <- read.csv("data-raw/optionA.csv")
optionA$opt <- "A"
optionB <- read.csv("data-raw/optionB.csv")
optionB$opt <- "B"
optionC <- read.csv("data-raw/optionC.csv")
optionC$opt <- "C"
optionD <- read.csv("data-raw/optionD.csv")
optionD$opt <- "D"
optionE <- read.csv("data-raw/optionE.csv")
optionE$opt <- "E"
optionF <- read.csv("data-raw/optionF.csv")
optionF$opt <- "F"
optionG <- read.csv("data-raw/optionG.csv")
optionG$opt <- "G"
optionH <- read.csv("data-raw/optionH.csv")
optionH$opt <- "H"

viridis.map <- rbind(optionA, optionB, optionC, optionD, optionE, optionF, optionG, optionH)

# usethis::use_data(viridis.map, overwrite = TRUE)

if (file.exists("R/zzz.R"))
  file.remove("R/zzz.R")

write("#' @export", file = "R/zzz.R")
write("viridis.map <- data.frame(", file = "R/zzz.R", append = TRUE)
write(paste0("\tR = c(", paste0(viridis.map$R, collapse = ", "), "),"), file = "R/zzz.R", append = TRUE)
write(paste0("\tG = c(", paste0(viridis.map$G, collapse = ", "), "),"), file = "R/zzz.R", append = TRUE)
write(paste0("\tB = c(", paste0(viridis.map$B, collapse = ", "), "),"), file = "R/zzz.R", append = TRUE)
write(paste0("\topt = c(", paste0("'", viridis.map$opt, "'", collapse = ", "), ")"), file = "R/zzz.R", append = TRUE)
write(")", file = "R/zzz.R", append = TRUE)


R/viridis.R
#' @title Color Map Data
#'
#' @description A data set containing the RGB values of the color maps included
#'  in the package. These are:
#'  \itemize{
#'   \item{}{'magma', 'inferno', 'plasma', and 'viridis' as defined in Matplotlib
#'    for Python. These color maps are designed in such a way that they will
#'    analytically be perfectly perceptually-uniform, both in regular form and
#'    also when converted to black-and-white. They are also designed to be
#'    perceived by readers with the most common form of color blindness. They
#'    were created by \href{https://github.com/stefanv}{Stéfan van der Walt}
#'    and \href{https://github.com/njsmith}{Nathaniel Smith};}
#'   \item{}{'cividis', a corrected version of 'viridis', 'cividis', developed by
#'    Jamie R. Nuñez, Christopher R. Anderton, and Ryan S. Renslow, and
#'    originally ported to R by Marco Sciaini. It is designed to be perceived by
#'    readers with all forms of color blindness;}
#'   \item{}{'rocket' and 'mako' as defined in Seaborn for Python;}
#'   \item{}{'turbo', an improved Jet rainbow color map for reducing false detail,
#'    banding and color blindness ambiguity.}
#'  }
#'
#' @references
#' \itemize{
#'   \item{}{'magma', 'inferno', 'plasma', and 'viridis': https://bids.github.io/colormap/}
#'   \item{}{'cividis': https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0199239}
#'   \item{}{'rocket' and 'mako': https://seaborn.pydata.org/index.html}
#'   \item{}{'turbo': https://ai.googleblog.com/2019/08/turbo-improved-rainbow-colormap-for.html}
#' }
#'
#' @format A data frame with 2048 rows and 4 variables:
#' \itemize{
#'   \item{R: }{Red value;}
#'   \item{G: }{Green value;}
#'   \item{B: }{Blue value;}
#'   \item{opt: }{The colormap "option" (A: magma; B: inferno; C: plasma;
#'    D: viridis; E: cividis; F: rocket; G: mako; H: turbo).}
#' }
#'
#' @author Simon Garnier: \email{garnier@@njit.edu} / \href{https://twitter.com/sjmgarnier}{@@sjmgarnier}
"viridis.map"


#' @title Viridis Color Palettes
#'
#' @description This function creates a vector of \code{n} equally spaced colors
#'  along the selected color map.
#'
#' @param n The number of colors (\eqn{\ge 1}) to be in the palette.
#'
#' @param alpha	The alpha transparency, a number in [0,1], see argument alpha in
#' \code{\link[grDevices]{hsv}}.
#'
#' @param begin The (corrected) hue in [0,1] at which the color map begins.
#'
#' @param end The (corrected) hue in [0,1] at which the color map ends.
#'
#' @param direction Sets the order of colors in the scale. If 1, the default,
#'  colors are ordered from darkest to lightest. If -1, the order of colors is
#'  reversed.
#'
#' @param option A character string indicating the color map option to use.
#'  Eight options are available:
#'  \itemize{
#'   \item{}{"magma" (or "A")}
#'   \item{}{"inferno" (or "B")}
#'   \item{}{"plasma" (or "C")}
#'   \item{}{"viridis" (or "D")}
#'   \item{}{"cividis" (or "E")}
#'   \item{}{"rocket" (or "F")}
#'   \item{}{"mako" (or "G")}
#'   \item{}{"turbo" (or "H")}
#'  }
#'
#' @return \code{viridis} returns a character vector, \code{cv}, of color hex
#'  codes. This can be used either to create a user-defined color palette for
#'  subsequent graphics by \code{palette(cv)}, a \code{col =} specification in
#'  graphics functions or in \code{par}.
#'
#' @author Simon Garnier: \email{garnier@@njit.edu} / \href{https://twitter.com/sjmgarnier}{@@sjmgarnier}
#'
#' @details
#' \if{html}{Here are the color scales:
#'   \figure{viridis-scales.png}{options: style="display:block;margin-left:auto;margin-right:auto;width:750px;max-width:75\%;"}
#' }
#' \if{latex}{Here are the color scales:
#'   \out{\begin{center}}\figure{viridis-scales.png}\out{\end{center}}
#' }
#'
#' \code{magma()}, \code{plasma()}, \code{inferno()}, \code{cividis()},
#'  \code{rocket()}, \code{mako()}, and \code{turbo()} are convenience functions
#'  for the other color map options, which are useful when the scale must be
#'  passed as a function name.
#'
#' Semi-transparent colors (\eqn{0 < alpha < 1}) are supported only on some
#'  devices: see \code{\link[grDevices]{rgb}}.
#'
#' @examples
#' library(ggplot2)
#' library(hexbin)
#'
#' dat <- data.frame(x = rnorm(10000), y = rnorm(10000))
#'
#' ggplot(dat, aes(x = x, y = y)) +
#'   geom_hex() + coord_fixed() +
#'   scale_fill_gradientn(colours = viridis(256, option = "D"))
#'
#' # using code from RColorBrewer to demo the palette
#' n = 200
#' image(
#'   1:n, 1, as.matrix(1:n),
#'   col = viridis(n, option = "D"),
#'   xlab = "viridis n", ylab = "", xaxt = "n", yaxt = "n", bty = "n"
#' )
#' @export
viridis <- function(n, alpha = 1, begin = 0, end = 1, direction = 1, option = "D") {

  if (begin < 0 | begin > 1 | end < 0 | end > 1) {
    stop("begin and end must be in [0,1]")
  }

  if (abs(direction) != 1) {
    stop("direction must be 1 or -1")
  }

  if (n == 0) {
    return(character(0))
  }

  if (direction == -1) {
    tmp <- begin
    begin <- end
    end <- tmp
  }

  option <- switch(EXPR = option,
                   A = "A", magma = "A",
                   B = "B", inferno = "B",
                   C = "C", plasma = "C",
                   D = "D", viridis = "D",
                   E = "E", cividis = "E",
                   F = "F", rocket = "F",
                   G = "G", mako = "G",
                   H = "H", turbo = "H",
                   {warning(paste0("Option '", option, "' does not exist. Defaulting to 'viridis'.")); "D"})

  map <- viridisLite::viridis.map[viridisLite::viridis.map$opt == option, ]
  map_cols <- grDevices::rgb(map$R, map$G, map$B)
  fn_cols <- grDevices::colorRamp(map_cols, space = "Lab", interpolate = "spline")
  cols <- fn_cols(seq(begin, end, length.out = n)) / 255
  grDevices::rgb(cols[, 1], cols[, 2], cols[, 3], alpha = alpha)
}


#' @rdname viridis
#'
#' @return \code{viridisMap} returns a \code{n} lines data frame containing the
#'  red (\code{R}), green (\code{G}), blue (\code{B}) and alpha (\code{alpha})
#'  channels of \code{n} equally spaced colors along the selected color map.
#'  \code{n = 256} by default.
#'
#' @export
viridisMap <- function(n = 256, alpha = 1, begin = 0, end = 1, direction = 1,
                       option = "D") { # nocov start
  if (begin < 0 | begin > 1 | end < 0 | end > 1) {
    stop("begin and end must be in [0,1]")
  }

  if (abs(direction) != 1) {
    stop("direction must be 1 or -1")
  }

  if (n == 0) {
    return(data.frame(R = double(0), G = double(0), B = double(0), alpha = double(0)))
  }

  if (direction == -1) {
    tmp <- begin
    begin <- end
    end <- tmp
  }

  option <- switch(EXPR = option,
                   A = "A", magma = "A",
                   B = "B", inferno = "B",
                   C = "C", plasma = "C",
                   D = "D", viridis = "D",
                   E = "E", cividis = "E",
                   E = "F", rocket = "F",
                   E = "G", mako = "G",
                   H = "H", turbo = "H",
                   {warning(paste0("Option '", option, "' does not exist. Defaulting to 'viridis'.")); "D"})

  map <- viridisLite::viridis.map[viridisLite::viridis.map$opt == option, ]
  map_cols <- grDevices::rgb(map$R, map$G, map$B)
  fn_cols <- grDevices::colorRamp(map_cols, space = "Lab", interpolate = "spline")
  cols <- fn_cols(seq(begin, end, length.out = n)) / 255
  data.frame(R = cols[, 1], G = cols[, 2], B = cols[, 3], alpha = alpha)
} # nocov end

#' @rdname viridis
#' @export
magma <- function(n, alpha = 1, begin = 0, end = 1, direction = 1) {
  viridis(n, alpha, begin, end, direction, option = "magma")
}

#' @rdname viridis
#' @export
inferno <- function(n, alpha = 1, begin = 0, end = 1, direction = 1) {
  viridis(n, alpha, begin, end, direction, option = "inferno")
}

#' @rdname viridis
#' @export
plasma <- function(n, alpha = 1, begin = 0, end = 1, direction = 1) {
  viridis(n, alpha, begin, end, direction, option = "plasma")
}

#' @rdname viridis
#' @export
cividis <- function(n, alpha = 1, begin = 0, end = 1, direction = 1) {
  viridis(n, alpha, begin, end, direction, option = "cividis")
}

#' @rdname viridis
#' @export
rocket <- function(n, alpha = 1, begin = 0, end = 1, direction = 1) {
  viridis(n, alpha, begin, end, direction, option = "rocket")
}

#' @rdname viridis
#' @export
mako <- function(n, alpha = 1, begin = 0, end = 1, direction = 1) {
  viridis(n, alpha, begin, end, direction, option = "mako")
}

#' @rdname viridis
#' @export
turbo <- function(n, alpha = 1, begin = 0, end = 1, direction = 1) {
  viridis(n, alpha, begin, end, direction, option = "turbo")
}
